File: src/main/java/alfheim/common/integration/minetweaker/handler/MTHandlerTradePortal.java
package alfheim.common.integration.minetweaker.handler;

import static alfheim.common.integration.minetweaker.MinetweakerAlfheimConfig.*;

import alfheim.api.AlfheimAPI;
import alfheim.api.ModInfo;
import minetweaker.IUndoableAction;
import minetweaker.MineTweakerAPI;
import minetweaker.api.item.IItemStack;
import net.minecraft.item.ItemStack;
import stanhebben.zenscript.annotations.ZenClass;
import stanhebben.zenscript.annotations.ZenMethod;

@ZenClass("mods." + ModInfo.MODID + ".TradePortal")
public class MTHandlerTradePortal {

	@ZenMethod
	public static void banRetrade(IItemStack output) {
		MineTweakerAPI.apply(new Ban(getStack(output)));
	}

	private static class Ban implements IUndoableAction {

		private final ItemStack output;
		
		public Ban(ItemStack stack) {
			output = stack;
		}

		@Override
		public void apply() {
			AlfheimAPI.banRetrade(output);
		}

		@Override
		public boolean canUndo() {
			return false;
		}

		@Override
		public void undo() {
			throw new IllegalArgumentException("Don't cheat!");
		}

		@Override
		public String describe() {
			return String.format("Removing %s from Alfheim trade portal", output.getUnlocalizedName());
		}

		@Override
		public String describeUndo() {
			throw new IllegalArgumentException("Don't cheat!");
		}

		@Override
		public Object getOverrideKey() {
			return null;
		}
	}
}

File: src/main/java/alfheim/common/integration/minetweaker/handler/MTHandlerTradePortal.kt
package alfheim.common.integration.minetweaker.handler

import alfheim.common.integration.minetweaker.MinetweakerAlfheimConfig.*

import alfheim.api.AlfheimAPI
import alfheim.api.ModInfo
import minetweaker.IUndoableAction
import minetweaker.MineTweakerAPI
import minetweaker.api.item.IItemStack
import net.minecraft.item.ItemStack
import stanhebben.zenscript.annotations.ZenClass
import stanhebben.zenscript.annotations.ZenMethod

@ZenClass("mods." + ModInfo.MODID + ".TradePortal")
object MTHandlerTradePortal {
	
	@ZenMethod
	fun banRetrade(output: IItemStack) {
		MineTweakerAPI.apply(Ban(getStack(output)))
	}
	
	private class Ban(private val output: ItemStack): IUndoableAction {
		
		override fun apply() {
			AlfheimAPI.banRetrade(output)
		}
		
		override fun canUndo(): Boolean {
			return false
		}
		
		override fun undo() {
			throw IllegalArgumentException("Don't cheat!")
		}
		
		override fun describe(): String {
			return String.format("Removing %s from Alfheim trade portal", output.unlocalizedName)
		}
		
		override fun describeUndo(): String {
			throw IllegalArgumentException("Don't cheat!")
		}
		
		override fun getOverrideKey(): Any? {
			return null
		}
	}
}
